### ISSUE TITLE
[BUG] Convergence issue in MLX when compared to PyTorch

### ISSUE THREAD

**Issue by mzy2240** [ml-explore/mlx#1324]
**Describe the bug**

The MLX model failed to converge after the second iteration while the PyTorch counterpart works fine.
![losses](https://github.com/user-attachments/assets/17da66fa-7459-4c51-aebb-06700e5486dc)
The plot is coming from the minimal reproducible example. In the actual use case, due to much greater amount of training data, the difference is much more obvious: the pytorch model losses reduce to 0.1 within 10 iterations, while the MLX model stuck at around 1.0 after the first iteration.

**To Reproduce**

[debug 2.ipynb.zip](https://github.com/user-attachments/files/16603797/debug.2.ipynb.zip)
[dataset.zip](https://github.com/user-attachments/files/16603799/dataset.zip)
To reproduce, I have attached the zipped notebook file. You only need to unzip and open it, no need to unzip the dataset.

**Expected behavior**
The training losses should keep the trend of decrease after the second iterations.

**Desktop (please complete the following information):**
 - OS Version: [e.g. MacOS 14.1.2] 13.6.6
 - Version [e.g. 0.7.0] 0.16.3

**Additional context**
Related to [discussion](https://github.com/ml-explore/mlx/discussions/1322)

**Comment by awni** [comment]
There's quite a few differences here between the two versions with some bugs as well. Let me delineate everything:

1. Everything in MLX is batch first. So your forward function is buggy right now. You should change it to the following (basically get rid of all the transposing):

```python
    def __call__(self, src, tgt):
        src = self.encode(src)
        out = self.decode(tgt, src)
        return out
    
    def encode(self, src):
        src_start = self.input_projection(src)
        batch_size, in_sequence_len = src_start.shape[:2]
        pos_encoder = mx.repeat(mx.arange(0, in_sequence_len).reshape((1,-1)), batch_size, axis=0)
        pos_encoder = self.input_pos_embedding(pos_encoder)
        src = src_start + pos_encoder
        src = self.encoder(src, mask=None)
        return src
    
    def decode(self, tgt, memory):
        tgt_start = self.output_projection(tgt)
        batch_size, out_sequence_len = tgt_start.shape[:2]
        pos_decoder = mx.repeat(mx.arange(0, out_sequence_len).reshape((1,-1)), batch_size, axis=0)
        pos_decoder = self.output_pos_embedding(pos_decoder)
        tgt = tgt_start + pos_decoder
        # tgt_mask = transformer_generate_tgt_mask(out_sequence_len)
        out = self.decoder(tgt, memory, None, None) + tgt_start
        out = self.output_linear_layer(out)
        return out
```

2. MLX uses an ubias-corrected Adam. This appears to be quite important for your use case to converge.. though I can't say why. You can roll your own Adam with the bias correction as follows:

```python
class Adam(optim.Optimizer):
    def __init__(
        self,
        learning_rate: Union[float, Callable[[mx.array], mx.array]],
        betas: List[float] = [0.9, 0.999],
        eps: float = 1e-8,
    ):
        super().__init__()
        
        self._maybe_schedule("learning_rate", learning_rate)
        self.betas = betas
        self.eps = eps
    
    def init_single(self, parameter: mx.array, state: dict):
        """Initialize optimizer state"""
        state["m"] = mx.zeros_like(parameter)
        state["v"] = mx.zeros_like(parameter)
        
    def apply_single(self, gradient: mx.array, parameter: mx.array, state: dict):
        """Performs the Adam parameter update and stores :math:`v` and
        :math:`m` in the optimizer state."""
        lr = self.learning_rate.astype(gradient.dtype)
        b1, b2 = self.betas
        eps = self.eps
        step = self.step
        b1corrected = 1 - b1**step
        b2corrected = 1 - b2**step
        
        m = state["m"]
        v = state["v"]
        m = b1 * m + (1 - b1) * gradient
        v = b2 * v + (1 - b2) * mx.square(gradient)
        state["m"] = m
        state["v"] = v

        lr = lr / b1corrected
        v = v / b2corrected

        return parameter - lr * m / (mx.sqrt(v) + eps)
```

3. There is actually a bug in our `norm_first=False` encoder layer (though this doesn't affect the results for you too much. I will send a patch.
4. The MLP dimensions default to `2048` in PyTorch but `4 * model_dim` in MLX which in your case will be `1024`. So the MLX model has less capacity. Though again this doesn't matter much for your case.

**Comment by mzy2240** [comment]
Wow really appreciate your answers! Glad to know this issue help improve the code a bit😜

**Comment by awni** [comment]
I think this is safe to close for now. Discussed with @angeloskath we will hold off on making a bias corrected Adam since it's not clearly a win (can be worse in some cases) and is slower.

Also one thing that helps a lot in your case even with an unbias corrected Adam is to use a warmup on the learning rate:

```python
optimizer = optim.Adam(learning_rate=optim.linear_schedule(0, 1e-3, 100))
```

Using that worked quite well actually and it should be more robust in general to the optimizer settings.

**Comment by mzy2240** [comment]
Sounds great! Looking forward to the new patch!
